File: tosho_mango/cli/impl/kmkc/download.py
"""
MIT License

Copyright (c) 2023-present noaione

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
"""

from __future__ import annotations

from pathlib import Path

import click
import msgspec
from requests import HTTPError

from tosho_mango import models, term
from tosho_mango.cli.base import ToshoMangoCommandHandler
from tosho_mango.sources.kmkc.config import KMConfigWeb
from tosho_mango.sources.kmkc.constants import CDN_HOST
from tosho_mango.sources.kmkc.dto import (
    EpisodeEntry,
    PremiumTicketInfo,
    TitleNode,
    TitleTicketInfo,
    UserPoint,
)
from tosho_mango.sources.kmkc.errors import KMAPIError, KMNotEnoughPointError
from tosho_mango.sources.kmkc.imaging import bytes_to_image, descramble_target

from .. import options
from .common import make_web_client, select_single_account

__all__ = (
    "kmkc_title_download",
    "kmkc_title_auto_download",
)
console = term.get_console()


def create_chapters_info(manga_detail: TitleNode, chapters: list[EpisodeEntry] | None = None) -> bytes:
    chapters_dump: list[models.ChapterDetailDump] = []
    if chapters:
        for chapter in chapters:
            ch_dump = models.ChapterDetailDump(id=chapter.episode_id, main_name=chapter.episode_name)
            ch_dump.timestamp = int(chapter.start_time_datetime().timestamp())
            chapters_dump.append(ch_dump)
    return msgspec.json.format(
        msgspec.json.encode(
            models.MangaDetailDump(manga_detail.title_name, manga_detail.author_text, chapters_dump),
        ),
        indent=4,
    )


def get_output_directory(
    output_dir: Path, title_id: int | str, chapter_id: int | str | None = None, *, skip_create: bool = False
):
    pathing = output_dir / str(title_id)
    if not skip_create:
        pathing.mkdir(parents=True, exist_ok=True)
    if chapter_id is not None:
        pathing = pathing / str(chapter_id)
        if not skip_create:
            pathing.mkdir(parents=True, exist_ok=True)
    return pathing


@click.command(
    name="download",
    help="Download manga chapter(s) for a title",
    cls=ToshoMangoCommandHandler,
)
@options.title_id
@options.output_dir
@click.option(
    "-c",
    "--chapter",
    "chapter_ids",
    type=options.COMMA_SEPARATED_NUMBER,
    help="Specify the chapter ID to purchase",
    default=None,
    required=False,
)
@click.option(
    "--show-all",
    "show_all",
    help="Show all chapters (including unavailable)",
    is_flag=True,
    default=False,
)
@click.option(
    "-ap",
    "--auto-purchase",
    "auto_purchase",
    help="Automatically purchase chapters, if needed",
    is_flag=True,
    default=False,
)
@options.account_id
def kmkc_title_download(
    title_id: int,
    output_dir: Path,
    chapter_ids: list[int] | None = None,
    show_all: bool = False,
    auto_purchase: bool = False,
    account_id: str | None = None,
):
    # Remove chapters dupe
    if chapter_ids is not None:
        chapter_ids = list(dict.fromkeys(chapter_ids))

    account = select_single_account(account_id)
    if account is None:
        console.warning("Aborted")
        return
    if not isinstance(account, KMConfigWeb):
        console.error("Only web account is supported for now!")
        return

    client = make_web_client(account=account)
    console.info(f"Getting user point for [highlight]{account.username}[/highlight]...")

    try:
        user_wallet = client.get_user_point().point
    except KMAPIError as exc:
        console.error(f"Failed to get user wallet: {exc}")
        return

    console.info(f"Getting title ID for [highlight]{title_id}[/highlight]...")
    try:
        result = client.get_title_list([title_id])[0]
    except KMAPIError as exc:
        console.error(f"Failed to get title ID: {exc}")
        return
    except IndexError:
        console.error(f"Failed to get title ID: Title ID {title_id} not found")
        return

    console.info(f"Fetching [highlight]{result.title_name}[/highlight] title ticket...")
    try:
        ticket_entry = client.get_title_ticket(title_id)
    except KMAPIError as exc:
        console.error(f"Failed to get title ticket information: {exc}")
        return

    chapters_info: list[EpisodeEntry] = []
    console.info(f"Fetching [highlight]{len(result.episode_id_list)}[/highlight] chapters information...")
    for episode_ids in client.chunk_episodes(result.episode_id_list):
        try:
            chapters_info.extend(client.get_episode_list(episode_ids))
        except KMAPIError as exc:
            console.error(f"Failed to get chapter information: {exc}")
            return

    if not chapters_info:
        console.error("No chapters found")
        return

    chapters_info.sort(key=lambda x: x.episode_id)

    if chapter_ids is None:
        select_choices = [
            term.ConsoleChoice(str(chapter.episode_id), chapter.episode_name)
            for chapter in chapters_info
            if chapter.available() and not show_all
        ]
        selected = console.select("Select chapter to download", select_choices)
        if not selected:
            console.warning("No chapter selected, aborting")
            return

        chapter_ids = list(map(lambda x: int(x.name), selected))

    console.info(f"Downloading {len(chapter_ids)} chapters...")
    # Precalculate chapters consumption
    ids_lists = [chapter.episode_id for chapter in chapters_info]
    actual_chapters: list[EpisodeEntry] = []
    _wallet_copy = msgspec.json.decode(msgspec.json.encode(user_wallet), type=UserPoint)
    for ch_id in chapter_ids:
        if ch_id not in ids_lists:
            console.warning(f"Episode ID [bold]{ch_id}[/bold] is not available, skipping")
            continue

        id_index = ids_lists.index(ch_id)
        chapter = chapters_info[id_index]
        if chapter.available():
            actual_chapters.append(chapter)
            continue

        should_purchase = auto_purchase
        if not auto_purchase:
            should_purchase = console.confirm(
                f"Chapter [highlight]{chapter.episode_name}[/highlight] ({chapter.episode_id}) need to be purchased "
                f"for {chapter.point}P, continue?",
            )

        if should_purchase:
            if chapter.ticketable():
                ticket: PremiumTicketInfo | TitleTicketInfo | None = None
                if ticket_entry.premium_available():
                    ticket = ticket_entry.ticket_info.premium_ticket_info
                    ticket_entry.subtract_premium()
                elif ticket_entry.title_available():
                    ticket = ticket_entry.ticket_info.title_ticket_info
                    ticket_entry.subtract_title()
                if ticket is not None:
                    console.info(f"Using ticket to purchase: [highlight]{chapter.episode_name}[/highlight]")
                    try:
                        client.claim_episode_with_ticket(chapter.episode_id, ticket)
                        actual_chapters.append(chapter)
                    except KMAPIError as exc:
                        console.warning(f"Failed to purchase chapter, ignoring: {exc}")

            if not _wallet_copy.can_purchase(chapter.point):
                console.warning(
                    f"Chapter [highlight]{chapter.episode_name}[/highlight] ([bold]{chapter.episode_id}[/bold]) is "
                    "not available for purchase, skipping",
                )
                warn_info = f"Need {chapter.point} point"
                if chapter.ticketable():
                    warn_info += " or ticket"
                console.warning(warn_info)
                continue

            try:
                user_wallet = client.claim_episode_with_point(chapter, user_wallet)
                _wallet_copy.subtract(chapter.point)
                _wallet_copy.add(chapter.bonus_point)
                actual_chapters.append(chapter)
            except KMNotEnoughPointError:
                console.warning(f"Not enough point to purchase chapter: [bold]{chapter.episode_name}[/bold]")
            except KMAPIError as exc:
                console.warning(f"Failed to purchase chapter, ignoring: {exc}")

    if not actual_chapters:
        console.warning("No chapters to download after filtering, aborting")
        return

    title_dir = get_output_directory(output_dir, title_id)
    (title_dir / "_info.json").write_bytes(create_chapters_info(result, chapters_info))

    for chapter in actual_chapters:
        console.info(f"  Downloading chapter [highlight]{chapter.episode_name}[/highlight] ({chapter.episode_id})...")
        try:
            viewer_info = client.get_chapter_viewer(chapter.episode_id)
        except KMAPIError as exc:
            console.warning(f"Failed to get chapter viewer information, ignoring: {exc}")
            continue

        console.log(f"    Has: {len(viewer_info.page_list)} pages")
        console.log(f"    Seed: {viewer_info.scramble_seed}")

        CH_IMGS = get_output_directory(output_dir, title_id, chapter.episode_id, skip_create=True)
        if CH_IMGS.exists() and len(list(CH_IMGS.glob("*.png"))) >= len(viewer_info.page_list):
            console.warning(
                f"   Chapter [bold]{chapter.episode_name}[/bold] ({chapter.episode_id}) already downloaded, skipping",
            )
            continue

        CH_IMGS.mkdir(parents=True, exist_ok=True)
        for idx, image in enumerate(viewer_info.page_list):
            IMG_TARGET = CH_IMGS / f"p{idx:03d}.png"
            console.info(f"   Downloading image [bold]{IMG_TARGET.name}[/bold]...")

            try:
                img_req = client.client.get(
                    image,
                    headers={
                        "Host": CDN_HOST,
                    },
                )
                img_req.raise_for_status()
            except HTTPError as exc:
                console.warning(f"    Failed to download image, stopping: {exc}")
                break

            img = bytes_to_image(img_req.content)
            img_descram = descramble_target(img, 4, viewer_info.scramble_seed)
            img_descram.save(IMG_TARGET)


@click.command(
    name="autodownload",
    help="Automatically/batch download manga chapter(s) for a title",
    cls=ToshoMangoCommandHandler,
)
@click.argument("title_id", type=int, metavar="TITLE_ID", required=True)
@options.output_dir
@click.option(
    "-nt",
    "--no-ticket",
    "opt_no_ticket",
    is_flag=True,
    default=False,
    help="Force to only download using point",
)
@click.option(
    "-np",
    "--no-purchase",
    "opt_no_purchase",
    is_flag=True,
    default=False,
    help="Do not purchase chapter if possible",
)
@click.option(
    "-sf",
    "--start-from",
    "start_from",
    type=click.IntRange(min=1),
    default=None,
    help="Start downloading from chapter ID",
)
@click.option(
    "-ef",
    "--end-until",
    "end_until",
    type=click.IntRange(min=1),
    default=None,
    help="Stop downloading until chapter ID",
)
@options.account_id
def kmkc_title_auto_download(
    title_id: int,
    output_dir: Path,
    opt_no_ticket: bool = False,
    opt_no_purchase: bool = False,
    start_from: int | None = None,
    end_until: int | None = None,
    account_id: str | None = None,
):
    if start_from is not None and end_until is not None and start_from > end_until:
        raise click.BadParameter("Start chapter ID cannot be greater than end chapter ID")

    account = select_single_account(account_id)
    if account is None:
        console.warning("Aborted")
        return
    if not isinstance(account, KMConfigWeb):
        console.error("Only web account is supported for now!")
        return

    client = make_web_client(account=account)

    console.info(f"Getting user point for [highlight]{account.username}[/highlight]...")
    try:
        user_wallet = client.get_user_point().point
    except KMAPIError as exc:
        console.error(f"Failed to get user wallet: {exc}")
        return

    console.info(f"Getting title information for ID [highlight]{title_id}[/highlight]...")
    try:
        results = client.get_title_list([title_id])
    except KMAPIError as exc:
        console.error(f"Failed to get title information: {exc}")
        return

    if not results:
        console.error("Unable to find title information.")
        return

    result = results[0]

    console.info(f"Fetching [highlight]{result.title_name}[/highlight] title ticket...")
    try:
        ticket_entry = client.get_title_ticket(title_id)
    except KMAPIError as exc:
        console.error(f"Failed to get title ticket information: {exc}")
        return

    chapters_info: list[EpisodeEntry] = []
    console.info(f"Fetching [highlight]{len(result.episode_id_list)}[/highlight] chapters information...")
    for episode_ids in client.chunk_episodes(result.episode_id_list):
        try:
            chapters_info.extend(client.get_episode_list(episode_ids))
        except KMAPIError as exc:
            console.error(f"Failed to get chapter information: {exc}")
            return

    actual_chapters: list[EpisodeEntry] = []
    _wallet_copy = msgspec.json.decode(msgspec.json.encode(user_wallet), type=UserPoint)
    console.info(f"Prepurchasing chapters for [highlight]{result.title_name}[/highlight]")
    for chapter in chapters_info:
        if chapter.available():
            actual_chapters.append(chapter)
            continue

        if start_from is not None and chapter.episode_id < start_from:
            console.warning(f" Skipping chapter {chapter.episode_name} ({chapter.episode_id}) due to start from option")
            continue

        if end_until is not None and chapter.episode_id > end_until:
            break

        if opt_no_purchase:
            continue

        if chapter.ticketable() and not opt_no_ticket:
            ticket: PremiumTicketInfo | TitleTicketInfo | None = None
            if ticket_entry.premium_available():
                ticket = ticket_entry.ticket_info.premium_ticket_info
                ticket_entry.subtract_premium()
            elif ticket_entry.title_available():
                ticket = ticket_entry.ticket_info.title_ticket_info
                ticket_entry.subtract_title()
            if ticket is not None:
                console.info(f"Using ticket to purchase: [highlight]{chapter.episode_name}[/highlight]")
                try:
                    client.claim_episode_with_ticket(chapter.episode_id, ticket)
                    actual_chapters.append(chapter)
                except KMAPIError as exc:
                    console.warning(f"Failed to purchase chapter, ignoring: {exc}")

        if not _wallet_copy.can_purchase(chapter.point):
            console.warning(
                f"Chapter [highlight]{chapter.episode_name}[/highlight] ([bold]{chapter.episode_id}[/bold]) is "
                "not available for purchase, skipping",
            )
            warn_info = f"Need {chapter.point} point"
            if chapter.ticketable():
                warn_info += " or ticket"
            console.warning(warn_info)
            continue

        try:
            user_wallet = client.claim_episode_with_point(chapter, user_wallet)
            _wallet_copy.subtract(chapter.point)
            _wallet_copy.add(chapter.bonus_point)
            actual_chapters.append(chapter)
        except KMNotEnoughPointError:
            console.warning(f"Not enough point to purchase chapter: [bold]{chapter.episode_name}[/bold]")
        except KMAPIError as exc:
            console.warning(f"Failed to purchase chapter, ignoring: {exc}")

    if not actual_chapters:
        console.warning("No chapters to download after filtering, aborting")
        return

    title_dir = get_output_directory(output_dir, title_id)
    (title_dir / "_info.json").write_bytes(create_chapters_info(result, chapters_info))

    for chapter in actual_chapters:
        console.info(f"  Downloading chapter [highlight]{chapter.episode_name}[/highlight] ({chapter.episode_id})...")
        try:
            viewer_info = client.get_chapter_viewer(chapter.episode_id)
        except KMAPIError as exc:
            console.warning(f"Failed to get chapter viewer information, ignoring: {exc}")
            continue

        console.log(f"    Has: {len(viewer_info.page_list)} pages")
        console.log(f"    Seed: {viewer_info.scramble_seed}")

        CH_IMGS = get_output_directory(output_dir, title_id, chapter.episode_id, skip_create=True)
        if CH_IMGS.exists():
            console.warning(
                f"   Chapter [bold]{chapter.episode_name}[/bold] ({chapter.episode_id}) already downloaded, skipping",
            )
            continue

        CH_IMGS.mkdir(parents=True, exist_ok=True)
        for idx, image in enumerate(viewer_info.page_list):
            IMG_TARGET = CH_IMGS / f"p{idx:03d}.png"
            console.info(f"   Downloading image [bold]{IMG_TARGET.name}[/bold]...")

            try:
                img_req = client.client.get(
                    image,
                    headers={
                        "Host": CDN_HOST,
                    },
                )
                img_req.raise_for_status()
            except HTTPError as exc:
                console.warning(f"    Failed to download image, stopping: {exc}")
                break

            img = bytes_to_image(img_req.content)
            img_descram = descramble_target(img, 4, viewer_info.scramble_seed)
            img_descram.save(IMG_TARGET)

File: tosho/src/impl/kmkc/download.rs
use std::path::{Path, PathBuf};

use color_print::cformat;
use tosho_kmkc::models::{EpisodeNode, EpisodeViewerResponse, TicketInfoType, TitleNode};

use crate::{
    cli::ExitCode,
    r#impl::models::{ChapterDetailDump, MangaDetailDump},
};

use super::common::{common_purchase_select, select_single_account};

#[derive(Clone, Debug, Default)]
pub(crate) struct KMDownloadCliConfig {
    /// Disable all input prompt (a.k.a `autodownload`)
    pub(crate) no_input: bool,
    pub(crate) auto_purchase: bool,
    pub(crate) show_all: bool,

    pub(crate) chapter_ids: Vec<usize>,
    /// The start chapter range.
    ///
    /// Used only when `no_input` is `true`.
    pub(crate) start_from: Option<i32>,
    /// The end chapter range.
    ///
    /// Used only when `no_input` is `true`.
    pub(crate) end_at: Option<i32>,

    pub(crate) no_ticket: bool,
    pub(crate) no_point: bool,
}

fn check_downloaded_image_count(image_dir: &PathBuf, extension: &str) -> Option<usize> {
    // check if dir exist
    if !image_dir.exists() {
        return None;
    }

    // check if dir is dir
    if !image_dir.is_dir() {
        return None;
    }

    // check how many .avif files in the dir
    let mut count = 0;
    for entry in std::fs::read_dir(image_dir).unwrap() {
        let entry = entry.unwrap();
        let path = entry.path();
        if path.is_file() && path.extension().unwrap() == extension {
            count += 1;
        }
    }

    Some(count)
}

fn create_chapters_info(title: &TitleNode, chapters: Vec<EpisodeNode>) -> MangaDetailDump {
    let mut dumped_chapters: Vec<ChapterDetailDump> = vec![];
    for chapter in chapters {
        dumped_chapters.push(ChapterDetailDump::from(chapter));
    }

    MangaDetailDump::new(title.title.clone(), title.author.clone(), dumped_chapters)
}

fn get_output_directory(
    output_dir: &Path,
    title_id: i32,
    chapter_id: Option<i32>,
    create_folder: bool,
) -> PathBuf {
    let mut pathing = output_dir.to_path_buf();
    pathing.push(title_id.to_string());

    if let Some(chapter_id) = chapter_id {
        pathing.push(chapter_id.to_string());
    }

    if create_folder {
        std::fs::create_dir_all(&pathing).unwrap();
    }

    pathing
}

pub(crate) async fn kmkc_download(
    title_id: i32,
    dl_config: KMDownloadCliConfig,
    account_id: Option<&str>,
    output_dir: PathBuf,
    console: &mut crate::term::Terminal,
) -> ExitCode {
    let account = select_single_account(account_id);

    if account.is_none() {
        console.warn("Aborted");
        return 1;
    }

    if let (Some(start), Some(end)) = (dl_config.start_from, dl_config.end_at) {
        if start > end {
            console.error("Start chapter is greater than end chapter!");
            return 1;
        }
    }

    let account = account.unwrap();
    let (results, title_detail, all_chapters, client, user_point) = common_purchase_select(
        title_id,
        &account,
        true,
        dl_config.show_all,
        dl_config.no_input,
        console,
    )
    .await;

    match (results, title_detail, user_point) {
        (Ok(results), Some(title_detail), Some(user_point)) => {
            let results: Vec<&EpisodeNode> = results
                .iter()
                .filter(|&ch| {
                    if dl_config.no_input {
                        // check if chapter id is in range
                        match (dl_config.start_from, dl_config.end_at) {
                            (Some(start), Some(end)) => {
                                // between start and end
                                ch.id >= start && ch.id <= end
                            }
                            (Some(start), None) => {
                                ch.id >= start // start to end
                            }
                            (None, Some(end)) => {
                                ch.id <= end // 0 to end
                            }
                            _ => true,
                        }
                    } else {
                        dl_config.chapter_ids.is_empty()
                            || dl_config.chapter_ids.contains(&(ch.id as usize))
                    }
                })
                .collect();

            if results.is_empty() {
                console.warn("No chapters after filtered by selected chapter ids");
                return 1;
            }

            let mut wallet_copy = user_point.point.point.clone();
            let mut ticket_entry = user_point.ticket.clone();
            console.info(&format!("Downloading {} chapters...", results.len()));
            let mut download_chapters = vec![];
            // let mut chapters_with_bonus = vec![];
            for chapter in results {
                if chapter.is_available() {
                    download_chapters.push(chapter);
                    continue;
                }

                let mut should_purchase = dl_config.auto_purchase;
                if !dl_config.auto_purchase && !dl_config.no_input {
                    let prompt = cformat!(
                        "Chapter <m,s>{}</> (<s>{}</>) need to be purchased for {}P, continue?",
                        chapter.title,
                        chapter.id,
                        chapter.point
                    );
                    should_purchase = console.confirm(Some(&prompt));
                }

                if should_purchase {
                    if chapter.is_ticketable() && !dl_config.no_ticket {
                        let mut ticket_info = None;
                        if ticket_entry.is_title_available() {
                            console.info(&cformat!(
                                "  Using title ticket to purchase chapter <m,s>{}</> (<s>{}</>)...",
                                chapter.title,
                                chapter.id
                            ));
                            ticket_info = Some(TicketInfoType::Title(
                                ticket_entry.info.title.clone().unwrap(),
                            ));
                            ticket_entry.subtract_title();
                        } else if ticket_entry.is_premium_available() {
                            console.info(&cformat!(
                                "  Using premium ticket to purchase chapter <m,s>{}</> (<s>{}</>)...",
                                chapter.title,
                                chapter.id
                            ));
                            ticket_info = Some(TicketInfoType::Premium(
                                ticket_entry.info.premium.clone().unwrap(),
                            ));
                            ticket_entry.subtract_premium();
                        }

                        if let Some(ticket) = ticket_info {
                            match client.claim_episode_with_ticket(chapter.id, &ticket).await {
                                Ok(_) => {
                                    download_chapters.push(chapter);
                                    // if chapter.bonus_point > 0 {
                                    //     chapters_with_bonus.push(chapter.id);
                                    // }
                                    continue;
                                }
                                Err(e) => {
                                    console.error(&format!(
                                        "   Failed to purchase chapter, ignoring: {}",
                                        e
                                    ));
                                }
                            }
                        }
                    }

                    if dl_config.no_point {
                        continue;
                    }

                    if !wallet_copy.can_purchase(chapter.point.try_into().unwrap_or(0)) {
                        console.warn(&cformat!(
                            "   Chapter <m,s>{}</> (<s>{}</>), is not available for purchase, skipping",
                            chapter.title,
                            chapter.id
                        ));
                        let mut warn_info = format!("    Need {} point", chapter.point);
                        if chapter.is_ticketable() {
                            warn_info += " or ticket";
                        }
                        console.warn(&warn_info);
                        continue;
                    }

                    console.info(&cformat!(
                        "  Purchasing chapter <m,s>{}</> (<s>{}</>) for {}P...",
                        chapter.title,
                        chapter.id,
                        chapter.point
                    ));
                    match client.claim_episode(chapter, &mut wallet_copy).await {
                        Ok(_) => {
                            download_chapters.push(chapter);
                            // if chapter.bonus_point > 0 {
                            //     chapters_with_bonus.push(chapter.id);
                            // }
                        }
                        Err(e) => {
                            console
                                .error(&format!("   Failed to purchase chapter, ignoring: {}", e));
                        }
                    }
                }
            }

            if download_chapters.is_empty() {
                console.warn("No chapters to download after filtering, aborting");
                return 1;
            }

            download_chapters.sort_by(|&a, &b| a.id.cmp(&b.id));

            let title_dir = get_output_directory(&output_dir, title_id, None, true);
            let dump_info = create_chapters_info(&title_detail, all_chapters);

            let title_dump_path = title_dir.join("_info.json");
            dump_info
                .dump(&title_dump_path)
                .expect("Failed to dump title info");

            for chapter in download_chapters {
                console.info(&cformat!(
                    "  Downloading chapter <m,s>{}</> ({})...",
                    chapter.title,
                    chapter.id
                ));

                let viewer_info = client.get_episode_viewer(chapter).await;

                if let Err(e) = viewer_info {
                    console.error(&format!("Failed to get viewer info, ignoring: {}", e));
                    continue;
                }

                let viewer_info = viewer_info.unwrap();
                let image_dir =
                    get_output_directory(&output_dir, title_id, Some(chapter.id), false);

                // precheck
                match &viewer_info {
                    EpisodeViewerResponse::Web(web) => {
                        if web.pages.is_empty() {
                            console.warn(&cformat!(
                                "   Chapter <m,s>{}</> (<s>{}</>) has no pages, skipping",
                                chapter.title,
                                chapter.id
                            ));
                            continue;
                        }

                        if let Some(count) = check_downloaded_image_count(&image_dir, "png") {
                            if count >= web.pages.len() {
                                console.warn(&cformat!(
                                    "   Chapter <m,s>{}</> (<s>{}</>) already downloaded, skipping",
                                    chapter.title,
                                    chapter.id
                                ));
                                continue;
                            }
                        }

                        if console.is_debug() {
                            console.log(&format!("    Seed: {}", web.scramble_seed));
                        }
                    }
                    EpisodeViewerResponse::Mobile(mobile) => {
                        if mobile.pages.is_empty() {
                            console.warn(&cformat!(
                                "   Chapter <m,s>{}</> (<s>{}</>) has no pages, skipping",
                                chapter.title,
                                chapter.id
                            ));
                            continue;
                        }

                        if let Some(count) = check_downloaded_image_count(&image_dir, "jpg") {
                            if count >= mobile.pages.len() {
                                console.warn(&cformat!(
                                    "   Chapter <m,s>{}</> (<s>{}</>) already downloaded, skipping",
                                    chapter.title,
                                    chapter.id
                                ));
                                continue;
                            }
                        }
                    }
                };

                // create dir
                std::fs::create_dir_all(&image_dir).unwrap();
                let image_blocks: Vec<tosho_kmkc::models::ImagePageNode> = match &viewer_info {
                    EpisodeViewerResponse::Mobile(mobile) => mobile.pages.clone(),
                    EpisodeViewerResponse::Web(web) => {
                        web.pages
                            .iter()
                            .map(|p| p.clone().into())
                            .collect::<Vec<tosho_kmkc::models::ImagePageNode>>()
                    }
                };
                let scramble_seed = match &viewer_info {
                    EpisodeViewerResponse::Mobile(_) => None,
                    EpisodeViewerResponse::Web(web) => Some(web.scramble_seed),
                };
                let force_extensions = match &viewer_info {
                    EpisodeViewerResponse::Mobile(_) => "jpg",
                    EpisodeViewerResponse::Web(_) => "png",
                };
                let total_image_count = image_blocks.len() as u64;
                for (idx, image) in image_blocks.iter().enumerate() {
                    let image_fn = format!("p{:03}.{}", idx, force_extensions);
                    let img_dl_path = image_dir.join(&image_fn);

                    let writer = tokio::fs::File::create(&img_dl_path)
                        .await
                        .expect("Failed to create image file!");

                    if console.is_debug() {
                        console.log(&cformat!(
                            "   Downloading image <s>{}</> to <s>{}</>...",
                            image.file_name(),
                            image_fn
                        ));
                    } else {
                        console.progress(total_image_count, 1, Some("Downloading".to_string()));
                    }

                    match client
                        .stream_download(&image.url, scramble_seed, writer)
                        .await
                    {
                        Ok(_) => {}
                        Err(err) => {
                            console.error(&format!("    Failed to download image: {}", err));
                            // silent delete the file
                            tokio::fs::remove_file(&img_dl_path)
                                .await
                                .unwrap_or_default();
                        }
                    }

                    // claim bonus point (disable for now :D)
                    // if chapters_with_bonus.contains(&chapter.id) {
                    //     console.info(&cformat!(
                    //         "   Claiming bonus point for chapter <m,s>{}</> (<s>{}</>)...",
                    //         chapter.title,
                    //         chapter.id
                    //     ));

                    //     match client.finish_episode_viewer(chapter).await {
                    //         Ok(finish_res) => {
                    //             console.info(&cformat!(
                    //                 "    Claimed <s,yellow>{}</> bonus point for chapter <m,s>{}</> (<s>{}</>)",
                    //                 finish_res.bonus_point,
                    //                 chapter.title,
                    //                 chapter.id
                    //             ));
                    //         }
                    //         Err(err) => {
                    //             console.error(&format!("    Failed to claim bonus point: {}", err));
                    //         }
                    //     }
                    // }
                }
                console.stop_progress(Some("Downloaded".to_string()));
            }

            0
        }
        _ => 1,
    }
}
